### PR TITLE
UndoRedo add version changed signal

### PR DESCRIPTION
**Pull request by AlexHolly**
added some functions to manage undo buttons

EDIT:
```
extends HBoxContainer

var undo_redo = UndoRedo.new()

var undo = Button.new()
var redo = Button.new()
var save = Button.new()

var do = Button.new()
var label = Label.new()
var saved_state = -1

func _ready():
	OS.set_window_size(Vector2(230, 80))
	undo.set_text("Undo")
	redo.set_text("Redo")
	save.set_text("Save")
	
	do.set_text("Do")
	label.set_text("0")
	
	add_child(undo)
	add_child(redo)
	add_child(save)
	
	add_child(do)
	add_child(label)
	
	undo_redo.connect("version_changed", self, "update_undo_button_state")
	
	undo.connect("button_down", undo_redo, "undo")
	redo.connect("button_down", undo_redo, "redo")
	save.connect("button_down", self, "save_state")
	do.connect("button_down", self, "do_changes")
	update_undo_button_state()

func update_undo_button_state():
	undo.set_disabled(!undo_redo.has_undo())
	redo.set_disabled(!undo_redo.has_redo())
	save.set_disabled(saved_state==undo_redo.get_version())

func save_state():
	save.set_disabled(true)
	saved_state = undo_redo.get_version()

func do_changes():
	undo_redo.create_action("do Change")
	undo_redo.add_undo_method(label, "set_text", str(undo_redo.get_version()-1))
	undo_redo.add_do_method(label, "set_text", str(undo_redo.get_version()))
	undo_redo.commit_action()
```

### CODE REVIEW

**Comment by akien-mga** [comment]
Sorry for the delay.

Could you be more specific about the use case so that we can assess whether it's a good addition to the API?

**Comment by akien-mga** [comment]
Alright, seems sensible. Could you rebase and add documentation for the new methods and signal?

**Review by akien-mga — APPROVED** [review]
OK to merge, pending rebase.

**Comment by akien-mga** [comment]
> Could you rebase and add documentation for the new methods and signal?

Bump.

File: doc/classes/UndoRedo.xml [hunk 1]
```diff
@@ -154,6 +169,12 @@
 				Redo last action.
 			</description>
 		</method>
+      <signals>
```
Comment (akien-mga) [inline doc/classes/UndoRedo.xml]: Indentation is wrong, you can run `godot --doctool .` to make sure that your template is properly formatted (should also fix the extra newline added at the end of the file).

File: doc/classes/UndoRedo.xml [hunk 2]
```diff
@@ -154,6 +169,12 @@
...
+		   <signal name="version_changed">
+			   <description>
```
Comment (akien-mga) [inline doc/classes/UndoRedo.xml]: This should be documented too.

**Comment by AlexHolly** [comment]
Any idea why `./bin/godot.x11.tools.64 --doctool .` is removing many descriptions?

File: doc/classes/UndoRedo.xml [hunk 3]
```diff
@@ -158,19 +172,21 @@
...
 	</methods>
+	<signals>
+		<signal name="version_changed">
+			<description>
+			</description>
+		</signal>
+	</signals>
 	<constants>
 		<constant name="MERGE_DISABLE" value="0" enum="MergeMode">
-			Makes [code]do[/code]/[code]undo[/code] operations stay in separate actions.
```
Comment (akien-mga) [inline doc/classes/UndoRedo.xml]: Those shouldn't be removed, something wrong happened.

File: doc/classes/UndoRedo.xml [hunk 4]
```diff
@@ -141,6 +141,20 @@
 				This is useful mostly to check if something changed from a saved version.
 			</description>
 		</method>
+		<method name="has_redo">
+			<return type="bool">
+			</return>
+			<description>
+				returns true if redo available.
```
Comment (akien-mga) [inline doc/classes/UndoRedo.xml]: Sentences start with a capital letter...
Please also style "true" as `[code]true[/code]` as done in the rest of the doc.

Comment (akien-mga) [reply doc/classes/UndoRedo.xml]: Maybe be a bit more verbose too, e.g. "Returns [code]true[/code] if a 'redo' action is available, which means that one or more recent actions have been undone and could be reapplied."

File: doc/classes/UndoRedo.xml [hunk 5]
```diff
@@ -141,6 +141,20 @@
...
+			</description>
+		</method>
+		<method name="has_undo">
+			<return type="bool">
+			</return>
+			<description>
+				returns true if undo available.
```
Comment (akien-mga) [inline doc/classes/UndoRedo.xml]: Same thing about capital letter and `[code]true[/code]`.

I'd suggest "Returns [code]true[/code] if an 'undo' action is available."

**Review by akien-mga — CHANGES_REQUESTED** [review]

**Comment by akien-mga** [comment]
> Any idea why ./bin/godot.x11.tools.64 --doctool . is removing many descriptions?

Did you compile Godot from this branch? It might be that you have a very old Godot build with no support for constant descriptions, which thus removes them.

**Comment by AlexHolly** [comment]
You are right something was too old, solved it with `git clean -fdx`.

**Review by akien-mga — APPROVED** [review]

**Comment by akien-mga** [comment]
Thanks!

**Pull request merged**